src/endpoints/patch-session.js

```diff
@@ -14,29 +15,160 @@ module.exports = cors(async (req, res) => {
+    const udtPatches = body.patch.filter(patch => !patch.path.includes('/samples'))
 
-    const latestVersion = result.version + 1
+    let newJSON = {...JSON.parse(session.udt_json)}
+    fjp.applyPatch(newJSON, udtPatches)
+
+    await patchSamples(session.short_id, samplePatches)
+    const latestVersion = session.version + 1
 
     const insertStmt = db.prepare(`
         INSERT INTO session_state
         (previous_session_state_id, short_id, user_name, patch, version, udt_json)
         VALUES (?, ?, ?, ?, ?, ?)`);
 
     insertStmt.run(
-        result.session_state_id,
+        session.session_state_id,
         sessionId,
         body.userName || null,
         JSON.stringify(body.patch),
         latestVersion,
         JSON.stringify(newJSON)
     );
 
+    const samplesQueryResults = db.prepare(`SELECT  *
+                           FROM latest_sample_state
+                           WHERE session_short_id = ?`).all(session.short_id);
+
+    const samples = []
+    samplesQueryResults.forEach(sample => {
+        const content = JSON.parse(sample.content)
+        const annotation = JSON.parse(sample.annotation)
+        samples.push(Object.assign({}, content, {annotation}))
+    })
+
+    newJSON.samples = samples
+
     return send(res, 200, {latestVersion, hashOfLatestState: hash(newJSON)})
+})
+
+const patchCreateSamples = (async (sessionId, samplesToAdd) => {
+    const samplesQueries = []
+    if (samplesToAdd.length) {
+        let sessionSampleId = -1
+        const latestSample = db.prepare('SELECT * FROM sample_state WHERE session_short_id = ? ORDER BY session_sample_id DESC LIMIT 1').get(sessionId);
+        if (latestSample) {
+            sessionSampleId = parseInt(latestSample.session_sample_id)
+        }
+
+        samplesToAdd.forEach((sample) => {
+            sessionSampleId += 1
+            samplesQueries.push(db.prepare('INSERT INTO sample_state (session_short_id, session_sample_id, content) VALUES (?, ?, ?)').run(sessionId, sessionSampleId, JSON.stringify(sample)));
+        })
+    }
+    return samplesQueries
+})
+
+const patchRemoveSamples = (async (sessionId, patches) => {
+    if (patches.length) {
+        await asyncForEach(patches, async (patch) => {
+            const pathArray = patch.path.split('/')
+            const indexToRemove = pathArray[2]
+
+            await db.prepare('DELETE FROM sample_state WHERE session_short_id = ? AND session_sample_id = ? ').run(sessionId, indexToRemove);
+            await db.prepare(`
+                INSERT INTO sample_state 
+                    (session_short_id, session_sample_id, version, content, summary, annotation) 
+                SELECT 
+                    session_short_id, session_sample_id - 1, version + 1, content, summary, annotation
+                FROM 
+                    latest_sample_state
+                WHERE 
+                    session_short_id = ? AND session_sample_id > ?`).run(sessionId, indexToRemove);
```

Comment:
why not use UPDATE? The INSERT and DELETE is pretty intense, considering we just need to devrement  a bunch of samples.

Also you need to create a `sample_index`, because every table must have an immutable `_id` field that is the name of the title + _id. So the new columns should be:

`sample_state_id` (never changes)
`sample_index`

and we should not have:

`session_sample_id`